gitindex/index.go

```diff
@@ -532,6 +518,225 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
+		}
+
+		for _, fp := range patch.FilePatches() {
+			fromFile, toFile := fp.Files()
+
+			if fromFile == nil {
+				// file added
+				file := fileKey{Path: toFile.Path(), ID: toFile.Hash()}
+				repos[file] = hackSharedBlobLocation
+				branchMap[file] = append(branchMap[file], branch.Name)
+				continue
+			}
+
+			changedOrDeletedFile := fileKey{Path: fromFile.Path(), ID: fromFile.Hash()}
+
+			// If the file is either modified or deleted, we need to add ALL versions
+			// of this file (across all branches) to the build
+			for b, currentTree := range branchToCurrentTree {
+				f, err := currentTree.File(changedOrDeletedFile.Path)
+				if err != nil {
+					// the file doesn't exist in this branch
+					if errors.Is(err, object.ErrFileNotFound) {
+						continue
+					}
+
+					return nil, nil, nil, nil, fmt.Errorf("getting hash for file %q in branch %q: %w", changedOrDeletedFile.Path, b, err)
+				}
+
+				file := fileKey{Path: changedOrDeletedFile.Path, ID: f.ID()}
+
+				repos[file] = hackSharedBlobLocation
+				branchMap[file] = append(branchMap[file], b)
+			}
+
+			changedOrDeletedPaths = append(changedOrDeletedPaths, changedOrDeletedFile.Path)
+		}
+	}
+
+	// we need to de-duplicate the branch map before returning it - it's possible for the same
+	// branch to have been added multiple times if a file has been modified across multiple commits
+
+	for file, branches := range branchMap {
+		sort.Strings(branches)
+		branchMap[file] = uniq(branches)
+	}
+
+	// we also need to de-duplicate the list of changed or deleted file paths, it's also possible to have duplicates
+	// for the same reasoning as above
+
+	sort.Strings(changedOrDeletedPaths)
+	changedOrDeletedPaths = uniq(changedOrDeletedPaths)
+
+	return repos, branchMap, nil, changedOrDeletedPaths, nil
+}
+
+func prepareNormalBuild(options Options, repository *git.Repository) (repos map[fileKey]BlobLocation, branchMap map[fileKey][]string, branchVersions map[string]map[string]plumbing.Hash, err error) {
+	var repoCache *RepoCache
+	if options.Submodules {
+		repoCache = NewRepoCache(options.RepoCacheDir)
+	}
+
+	// branch => (path, sha1) => repo.
+	repos = map[fileKey]BlobLocation{}
+
+	// fileKey => branches
+	branchMap = map[fileKey][]string{}
+
+	// Branch => Repo => SHA1
+	branchVersions = map[string]map[string]plumbing.Hash{}
+
+	branches, err := expandBranches(repository, options.Branches, options.BranchPrefix)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("expandBranches: %w", err)
+	}
+
+	for _, b := range branches {
+		commit, err := getCommit(repository, options.BranchPrefix, b)
+		if err != nil {
+			if options.AllowMissingBranch && err.Error() == "reference not found" {
+				continue
+			}
+
+			return nil, nil, nil, fmt.Errorf("getCommit: %w", err)
+		}
+
+		tree, err := commit.Tree()
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("commit.Tree: %w", err)
+		}
+
+		ig, err := newIgnoreMatcher(tree)
```

Comment:
you aren't doing this in the delta version? we have an ignore file in the tree which can change what we index. ie we may now start ignoring a file. or when you do the diff you will have a file you should ignore / etc.

Reply:
I decided that dealing with the ignore file could be something that we do after the prototype. There seemed to be some small subtleties about filetombstones and `.sourcegraph/ignore` files interact. It's totally doable, but this seemed like something I could skip for now.

Reply:
Ok, lets just make sure we don't forget about that.

Reply:
https://github.com/sourcegraph/sourcegraph/issues/32922